# Algoritmos de optimización - Seminario

#### Nico Toscano

#### URL: https://github.com/NicolaiOcin/03MAIR---Algoritmos-de-Optimizacion---2019/tree/master/SEM          

In [1]:
# Librerias
import random
import copy
import itertools
from time import time
from functools import wraps

def calcular_tiempo(f):
    @wraps(f)
    def cronometro(*args, **kwargs):
        t_inicial = time()
        salida = f(*args, **kwargs)
        t_final = time()
        print('Tiempo transcurrido (en segundos): {}'.format(t_final - t_inicial))
        return salida
    return cronometro

In [11]:
# Equipos
Equipos = {'A':['Eq1','Eq2','Eq3','Eq4'], 'B':['Eq5','Eq6','Eq7','Eq8','Eq9','Eq10','Eq11','Eq12','Eq13','Eq14'], 
      'C':['Eq15','Eq16','Eq17','Eq18','Eq19','Eq20']}
Equipos_lista = ['Eq1','Eq2','Eq3','Eq4','Eq5','Eq6','Eq7','Eq8','Eq9','Eq10','Eq11','Eq12','Eq13','Eq14','Eq15',
                'Eq16','Eq17','Eq18','Eq19','Eq20']

In [14]:
# Ejemplo de un resultado de una jornada (colacación de los partidos en los distintos horarios posibles) 
Results = { 'V20':['Eq1','Eq2'],
            'S12':['Eq3','Eq4'], 
            'S16':['Eq5','Eq6'],
            'S18':['Eq7','Eq8'],
            'S20':['Eq9','Eq10'],
            'D12':['Eq11','Eq12'],
            'D16':['Eq13','Eq14'], 
            'D18':['Eq15','Eq16'],
            'D20':['Eq17','Eq18'],
            'L20':['Eq19','Eq20']}

In [21]:
# Claves del diccionario de Results (todos los posibles horarios)
key_names = list(Results.keys())
key_names

['V20', 'S12', 'S16', 'S18', 'S20', 'D12', 'D16', 'D18', 'D20', 'L20']

In [16]:
# Creación de una jornada aleatoria
def generar_jornada():
    j = []
    Equipos_lista_copy = copy.deepcopy(Equipos_lista)
    for i in range(10):
        x = random.sample(Equipos_lista_copy,2)
        j.append(x)
        Equipos_lista_copy.remove(x[0])
        Equipos_lista_copy.remove(x[1])
    return j

In [39]:
# Ejemplo de una jornada aleatoria
j = generar_jornada()
j

[['Eq19', 'Eq9'],
 ['Eq8', 'Eq5'],
 ['Eq15', 'Eq17'],
 ['Eq6', 'Eq1'],
 ['Eq7', 'Eq20'],
 ['Eq12', 'Eq4'],
 ['Eq16', 'Eq10'],
 ['Eq3', 'Eq18'],
 ['Eq11', 'Eq13'],
 ['Eq2', 'Eq14']]

In [22]:
# Para pasar de lista a diccionario (le pasamos la lista y las claves del diccionario)
def dic(sol,keys):
    r = {}
    d = len(sol)-len(keys)
    if d>0:
        key_names_copy =copy.deepcopy(key_names)
        for i in key_names_copy:
            if i in keys:
                key_names_copy.remove(i)
        keys = keys + key_names_copy
        
    for i in range(len(sol)):
        if i == (len(sol)-1) and keys[i] != 'L20' and 'L20' not in keys:
            r.update( {'L20' : sol[i]} )
        else:
            r.update( {keys[i] : sol[i]} )
        
    return r

In [40]:
# Ejemplo de pasar la lista de la jornada (que hemos generado aleatoriamente) como solucion misma, y la pasamos a diccionario
dic_ej = dic(j,key_names)
dic_ej

{'V20': ['Eq19', 'Eq9'],
 'S12': ['Eq8', 'Eq5'],
 'S16': ['Eq15', 'Eq17'],
 'S18': ['Eq6', 'Eq1'],
 'S20': ['Eq7', 'Eq20'],
 'D12': ['Eq12', 'Eq4'],
 'D16': ['Eq16', 'Eq10'],
 'D18': ['Eq3', 'Eq18'],
 'D20': ['Eq11', 'Eq13'],
 'L20': ['Eq2', 'Eq14']}

In [23]:
# Pasar de un diccionario a una lista
def dic_inv(sol):
    kn = list(sol.keys())
    r = []
    for i in range(len(kn)):
        r.append(sol[kn[i]])
    return r

In [41]:
# Ejemplo de pasar la solución diccionario (de la jornada generada aleatoriamente) a lista
dic_inv_ej = dic_inv(dic_ej)
dic_inv_ej

[['Eq19', 'Eq9'],
 ['Eq8', 'Eq5'],
 ['Eq15', 'Eq17'],
 ['Eq6', 'Eq1'],
 ['Eq7', 'Eq20'],
 ['Eq12', 'Eq4'],
 ['Eq16', 'Eq10'],
 ['Eq3', 'Eq18'],
 ['Eq11', 'Eq13'],
 ['Eq2', 'Eq14']]

In [44]:
# Creación de una solución aleatoria sin coincidencias de partidos (le pasamos una jornada)
def crear_solucion(lequipos):
    r = {}
    equipos_s = copy.deepcopy(lequipos)
    for i in key_names:
        x = random.choice(equipos_s)
        r.update( {i : x} )
        equipos_s.remove(x)
    return r

In [48]:
# Ejemplo de solucion sin coincidencias de partidos
solucion_ej = crear_solucion(j)
solucion_ej

{'V20': ['Eq11', 'Eq13'],
 'S12': ['Eq16', 'Eq10'],
 'S16': ['Eq12', 'Eq4'],
 'S18': ['Eq19', 'Eq9'],
 'S20': ['Eq15', 'Eq17'],
 'D12': ['Eq8', 'Eq5'],
 'D16': ['Eq7', 'Eq20'],
 'D18': ['Eq3', 'Eq18'],
 'D20': ['Eq2', 'Eq14'],
 'L20': ['Eq6', 'Eq1']}

In [71]:
# Creación de una solución aleatoria con coincidencias de partidos (le pasamos una jornada). Siempre asignando un partido
# a V20 y L20 como se indica en el problema
def crear_solucion_c(jornada):
    r = {}
    jornada_copy = copy.deepcopy(jornada)
    for i in range(len(jornada)):
        if len(jornada_copy)<=2 and 'L20' not in r:
            x = jornada_copy
            if len(x)==1:
                r.update( {'L20' : x[0]} )
            else:
                r.update( {'L20' : x} )
            break
        if len(jornada_copy)>2:
            if random.getrandbits(1):
                val_r = random.randint(2,(len(jornada_copy)-1))
                x = random.sample(jornada_copy,val_r)
                r.update( {key_names[i] : x} )
                for j in x:
                    jornada_copy.remove(j)
            else:
                x = random.choice(jornada_copy)
                r.update( {key_names[i] : x} )
                jornada_copy.remove(x)
    return r

In [72]:
# Ejemplo de solucion sin coincidencias de partidos
solucion_c_ej = crear_solucion_c(j)
solucion_c_ej

{'V20': ['Eq8', 'Eq5'],
 'S12': [['Eq7', 'Eq20'], ['Eq3', 'Eq18'], ['Eq15', 'Eq17'], ['Eq16', 'Eq10']],
 'S16': ['Eq2', 'Eq14'],
 'S18': ['Eq12', 'Eq4'],
 'S20': ['Eq11', 'Eq13'],
 'L20': [['Eq19', 'Eq9'], ['Eq6', 'Eq1']]}

In [43]:
# Cantidad de partidos que hay en una solución (de un diccionario). Esto es para comprobar que haya el número de partidos
# correcto en la soluciones que proponemos (10 partidos)
def tamaño(results):
    a = dic_inv(results)
    suma = 0
    for i in a:
        for j in i:
            if j[0]=='E':
                suma += 1
                break
            else:
                suma +=1
    return suma

In [77]:
# Ejemplo cantidad de partidos en las soluciones aleatorias
t_ej_1 = tamaño(solucion_ej) # solución sin coincidencias
t_ej_2 = tamaño(solucion_c_ej) # solución con coincidencias
print("·Número partidos solución sin coincidencias:",t_ej_1)
print("·Número partidos solución con coincidencias:",t_ej_2)

·Número partidos solución sin coincidencias: 10
·Número partidos solución con coincidencias: 10


In [28]:
# Lista de partidos en una solucion (de un diccionario)
def partidos_lista(results):
    kn = list(results.keys())
    elem = []
    for i in kn:
        for j in results[i]:
            if j[0] == 'E':
                elem.append(results[i])
                break
            else:
                elem.append(j)
    return elem

In [127]:
# Ejemplo de partidos en las soluciones aleatorias
partidos_ej_1 = partidos_lista(solucion_ej) # solución sin coincidencias
partidos_ej_2 = partidos_lista(solucion_c_ej) # solución con coincidencias
print("·Lista partidos solución sin coincidencias: \n")
partidos_ej_1

·Lista partidos solución sin coincidencias: 



[['Eq11', 'Eq13'],
 ['Eq16', 'Eq10'],
 ['Eq12', 'Eq4'],
 ['Eq19', 'Eq9'],
 ['Eq15', 'Eq17'],
 ['Eq8', 'Eq5'],
 ['Eq7', 'Eq20'],
 ['Eq3', 'Eq18'],
 ['Eq2', 'Eq14'],
 ['Eq6', 'Eq1']]

In [129]:
print("·Lista partidos solución con coincidencias:")
partidos_ej_2

·Lista partidos solución con coincidencias:


[['Eq8', 'Eq5'],
 ['Eq7', 'Eq20'],
 ['Eq3', 'Eq18'],
 ['Eq15', 'Eq17'],
 ['Eq16', 'Eq10'],
 ['Eq2', 'Eq14'],
 ['Eq12', 'Eq4'],
 ['Eq11', 'Eq13'],
 ['Eq19', 'Eq9'],
 ['Eq6', 'Eq1']]

In [80]:
# Para juntar 2 diccionarios
def merge_two_dicts(x, y):
    z = x.copy()   
    z.update(y)    
    return z

In [94]:
# Comprobación de si el resultado es correcto segun las restricciones
def comprobacion(results):
    a = False
    elem = []
    c = dic_inv(results)
    for i in c:
        if len(i)==0:
            return True
        for j in i:
            if len(j)>1 and len(j)<=2:
                if (j[0] or j[1]) in elem:
                    a = True
        
            else:
                if j in elem:
                    a = True
        
            if len(j)>1 and len(j)<=2:
                elem.append(j[0]) 
                elem.append(j[1]) 
            else:
                elem.append(j)
    t = tamaño(results)

    if a:
        return True
    if a==False:
        return False    

In [101]:
# Ejemplo comprobación de las soluciones aleatorias
comprobacion_ej_1 = comprobacion(solucion_ej) # solución sin coincidencias
comprobacion_ej_2 = comprobacion(solucion_c_ej) # solución con coincidencias
print("·Comprobación solución sin coincidencias:",comprobacion_ej_1)
print("·Comprobación solución con coincidencias:",comprobacion_ej_2)
print("Nota: falso es que la solución es correcta")

·Comprobación solución sin coincidencias: False
·Comprobación solución con coincidencias: False
Nota: falso es que la solución es correcta


In [109]:
# Coeficientes segun las categorias de equipos
def cat(E1,E2):
    if E1 in Equipos['A'] and E2 in Equipos['A']:
        return 2
    if E1 in Equipos['A'] and E2 in Equipos['B']:
        return 1.3
    if E1 in Equipos['A'] and E2 in Equipos['C']:
        return 1
    if E1 in Equipos['B'] and E2 in Equipos['A']:
        return 1.3 
    if E1 in Equipos['B'] and E2 in Equipos['B']:
        return 0.9  
    if E1 in Equipos['B'] and E2 in Equipos['C']:
        return 0.75
    if E1 in Equipos['C'] and E2 in Equipos['A']:
        return 1
    if E1 in Equipos['C'] and E2 in Equipos['B']:
        return 0.75
    if E1 in Equipos['C'] and E2 in Equipos['C']:
        return 0.47  

In [110]:
# Coeficientes segun el horario que se asigna el partido
def coef(results):
    kn = list(results.keys())
    if 'V20' in kn:
        results.update( {'V20' : [x * 0.4 for x in results['V20']]} )
    if 'S12' in kn:
        results.update( {'S12' : [x * 0.55 for x in results['S12']]} ) 
    if 'S16' in kn:
        results.update( {'S16' : [x * 0.7 for x in results['S16']]} )    
    if 'S18' in kn:
         results.update( {'S18' : [x * 0.8 for x in results['S18']]} )
    if 'S20' in kn:
         results.update( {'S20' : [x * 1 for x in results['S20']]} )
    if 'D12' in kn:
         results.update( {'D12' : [x * 0.45 for x in results['D12']]} )   
    if 'D16' in kn:
         results.update( {'D16' : [x * 0.75 for x in results['D16']]} )
    if 'D18' in kn:
         results.update( {'D18' : [x * 0.5 for x in results['D18']]} )
    if 'D20' in kn:
         results.update( {'D20' : [x * 1 for x in results['D20']]} )
    if 'L20' in kn:
         results.update( {'L20' : [x * 0.4 for x in results['L20']]} )
    return results

In [111]:
# Coeficientes segun si hay concidencias en un horario
def coef_c(results):
    kn = list(results.keys())
    for i in kn:
        d = len(results[i])
       
        if d == 1:
             c = 1
        else:
            if d == 2:
                c = 1-0.25
            if d == 3:
                c = 1-0.45
            if d == 4:
                c = 1-0.6
            if d == 5:
                c = 1-0.7
            if d == 6:
                c = 1-0.75
            if d == 7:
                c = 1-0.78
            if d >= 8:
                c = 1-0.8
        results[i] = [x * c for x in results[i] ]
    return results

In [112]:
# Sumar los elementos de una lista
def suma(results):
    a = dic_inv(results)
    suma = 0
    for i in a:
        for j in i:
            suma += j
    return suma

In [115]:
# Valor de una solución (espectadores)
def valor(results):
    r = {}
    kn = list(results.keys())
    for i in kn:
        if results[i][0][0] == 'E':
            temp = cat(results[i][0], results[i][1])
            r.update( {i : [temp]} )
        else:
            temp_list = []
            for j in results[i]:
                temp = cat(j[0], j[1])
                temp_list.append(temp)
            r.update( {i : temp_list} )
    r2 = coef(r)
    r3 = coef_c(r2)
    val = suma(r3)
    return round(val,3)

In [116]:
# Ejemplo de calcular el valor de las soluciones aleatorias
valor_ej_1 = valor(solucion_ej) # solución sin coincidencias
valor_ej_2 = valor(solucion_c_ej) # solución con coincidencias
print("·Valor solución sin coincidencias:",valor_ej_1)
print("·Valor solución con coincidencias:",valor_ej_2)

·Valor solución sin coincidencias: 6.04
·Valor solución con coincidencias: 4.478


In [130]:
# Busqueda aleatoria de la mejor solución de un conjunto de soluciones (le pasamos la cantidad de soluciones que
# queremos que nos genere (N), la jornada, y "c" hace referencia al tipo de soluciones, con coincidencias (True) o sin (False))
def busqueda_aleatoria(N,jornada,c):
    mejor_solucion = {}
    mejor_solucion_valor = 0
    
    if c:
        s = crear_solucion_c
    else:
        s = crear_solucion
        
    for i in range(N):
        solucion = s(jornada)
        solucion_valor = valor(solucion)
        
        if solucion_valor > mejor_solucion_valor:
            mejor_solucion = solucion
            mejor_solucion_valor = solucion_valor    
    return mejor_solucion, mejor_solucion_valor

In [120]:
# Ejemplo de mejor solución dentro de 10000 soluciones
busqueda_aleat_ej_1 = busqueda_aleatoria(10000,j,False) # solución sin coincidencias
busqueda_aleat_ej_2 = busqueda_aleatoria(10000,j,True) # solución con coincidencias
print("·Mejor solución sin coincidencias:")
busqueda_aleat_ej_1

·Mejor solución sin coincidencias:


({'V20': ['Eq19', 'Eq9'],
  'S12': ['Eq8', 'Eq5'],
  'S16': ['Eq11', 'Eq13'],
  'S18': ['Eq2', 'Eq14'],
  'S20': ['Eq6', 'Eq1'],
  'D12': ['Eq7', 'Eq20'],
  'D16': ['Eq3', 'Eq18'],
  'D18': ['Eq16', 'Eq10'],
  'D20': ['Eq12', 'Eq4'],
  'L20': ['Eq15', 'Eq17']},
 6.716)

In [122]:
print("·Mejor solución con coincidencias:")
busqueda_aleat_ej_2

·Mejor solución con coincidencias:


({'V20': ['Eq7', 'Eq20'],
  'S12': ['Eq8', 'Eq5'],
  'S16': ['Eq12', 'Eq4'],
  'S18': ['Eq2', 'Eq14'],
  'S20': [['Eq3', 'Eq18'], ['Eq11', 'Eq13']],
  'D12': ['Eq15', 'Eq17'],
  'D16': [['Eq19', 'Eq9'], ['Eq6', 'Eq1']],
  'L20': ['Eq16', 'Eq10']},
 5.835)

In [173]:
# Generación de una población de soluciones. Le pasamos la jornada a partir de la cual queremos que nos encuentre, 
# el número de soluciones N, las soluciones podran ser con coincidencias o sin coincidencias
def generar_poblacion(jornada,N):
    return [random.choice([crear_solucion,crear_solucion_c])(jornada) for _ in range(N)]

In [195]:
# Ejemplo de una población de 5 soluciones, con coincidencias y sin concidencias en los horarios
poblacion_ej = generar_poblacion(j,2)
poblacion_ej

[{'V20': [['Eq3', 'Eq18'], ['Eq11', 'Eq13']],
  'S12': ['Eq8', 'Eq5'],
  'S16': [['Eq7', 'Eq20'], ['Eq16', 'Eq10']],
  'S18': [['Eq15', 'Eq17'], ['Eq6', 'Eq1']],
  'S20': ['Eq12', 'Eq4'],
  'L20': [['Eq19', 'Eq9'], ['Eq2', 'Eq14']]},
 {'V20': ['Eq8', 'Eq5'],
  'S12': ['Eq2', 'Eq14'],
  'S16': ['Eq7', 'Eq20'],
  'S18': ['Eq16', 'Eq10'],
  'S20': ['Eq19', 'Eq9'],
  'D12': ['Eq11', 'Eq13'],
  'D16': ['Eq15', 'Eq17'],
  'D18': ['Eq6', 'Eq1'],
  'D20': ['Eq3', 'Eq18'],
  'L20': ['Eq12', 'Eq4']}]

In [157]:
# Nos devuelve la mejor soución con su valor de una población
def evaluar_poblacion(pobl):
    mejor_solucion = {}
    mejor_solucion_valor = 0
    for p in pobl:
        solucion_valor = valor(p)
        if solucion_valor > mejor_solucion_valor:
            mejor_solucion = p
            mejor_solucion_valor = solucion_valor    
    return mejor_solucion, mejor_solucion_valor

In [196]:
# Ejemplo de evaluación de una población 
eval_poblacion_ej = evaluar_poblacion(poblacion_ej)
eval_poblacion_ej

({'V20': ['Eq8', 'Eq5'],
  'S12': ['Eq2', 'Eq14'],
  'S16': ['Eq7', 'Eq20'],
  'S18': ['Eq16', 'Eq10'],
  'S20': ['Eq19', 'Eq9'],
  'D12': ['Eq11', 'Eq13'],
  'D16': ['Eq15', 'Eq17'],
  'D18': ['Eq6', 'Eq1'],
  'D20': ['Eq3', 'Eq18'],
  'L20': ['Eq12', 'Eq4']},
 5.878)

In [177]:
# De la población escogemos dos soluciones (padres) y las combinamos (cruzamos los padres) entre ellas para que nos generen
# otras dos soluciones distintas (hijos)
def cruzar(poblacion,mutacion,jornada):
    poblacion_copia = copy.deepcopy(poblacion)
    poblacion_final =[]
    
    while len(poblacion_copia)>1:
        padre1,padre2 = random.sample(poblacion_copia   ,2)
        poblacion_copia.remove(padre1)
        poblacion_copia.remove(padre2)
        poblacion_final.extend(descendencia([padre1,padre2],mutacion,jornada))
    return poblacion_final     

In [185]:
# Parte de los padres que destinamos a cada hijo
# El hijo 1 se queda con los siguientes partidos:
# - partidos V20 y L20 del padre 1 y los demás partidos del padre 2
# El hijo 2 se queda con los siguientes partidos:
# - partidos V20 y L20 del padre 2 y los demás partidos del padre 1
def descendencia(padres,mutacion,jornada):
    kn1 = list(padres[0].keys())
    kn2 = list(padres[1].keys())
    hijo1 = {}
    hijo2 = {}
    
    hijo1.update({'V20': padres[0]['V20']})
    for i in kn2:
        if tamaño(hijo1)<9:
            if i not in hijo1 and i != 'L20':
                x = padres[1][i]
                hijo1.update({i: x})
    hijo1.update({'L20': padres[0]['L20']})
    
    
    hijo2.update({'V20': padres[1]['V20']})
    for i in kn1:
        if tamaño(hijo2)<9:
            if i not in hijo2 and i != 'L20':
                x = padres[0][i]
                hijo2.update({i: x})
    hijo2.update({'L20': padres[1]['L20']})

    hijo1 = factibilizar(jornada,hijo1)
    hijo2 = factibilizar(jornada,hijo2)
    
    
    return mutar(hijo1,mutacion),mutar(hijo2,mutacion)

In [180]:
# Eliminamos los elementos de una lista en función de si estan en otra lista (e)
def eliminar(results,e):
    for i in range(len(results)):
        for j in range(len(results[i])):
            if results[i][j] == e:
                results[i].pop(j)
                return results
            if results[i][j][0]=='E':
                if results[i] ==e:
                    results.pop(i)
                    return results
                break
    return results

In [182]:
# Eliminos los elementos repetidos de una misma lista
def eliminar_repetidos(results):
    kn = list(results.keys())
    elem = []
    c = dic_inv(results)
    for i in c:
        aux = False
        for j in i:
            if aux:
                break
            if j[0]=='E':
                if i in elem:
                    c = eliminar(c,i)
                else:
                    elem.append(i)
                aux = True      
            if j in elem:
                c = eliminar(c,j)
            if j not in elem and j[0] != 'E':
                elem.append(j)
    return dic(c,kn)

In [184]:
# Cuando cruzamos los padres, nos pueden salir partidos repetidos, así que tenemos que modificar la solución
# que se nos genera a la hora de cruzar los padres, para que sea factible
def factibilizar(jornada,solucion):
    r =solucion
    r = eliminar_repetidos(r)
    partidos_solucion = partidos_lista(r)
    partidos_desaparecidos = [] 
    kns = list(r.keys())
    for i in jornada:
        if i not in partidos_solucion:
            partidos_desaparecidos.append(i)  
    for i in range(len(jornada)):
        if len(partidos_desaparecidos)>0:
            if key_names[i] not in kns:
                if random.getrandbits(1) and len(partidos_desaparecidos)>1:
                    val_r = random.randint(2,(len(partidos_desaparecidos)))
                    x = random.sample(partidos_desaparecidos,val_r)
                    r.update( {key_names[i] : x} )
                    for j in x:
                        partidos_desaparecidos.remove(j)
                else:
                    x = random.choice(partidos_desaparecidos)
                    r.update( {key_names[i] : x} )
                    partidos_desaparecidos.remove(x)
            kns.append(key_names[i])
        else:
            break
    if comprobacion(r):
        r= crear_solucion_c(jornada)  
    l = len(r)-1
    kns = list(r.keys()) 
    if 'L20' not in kns:
        x = kns[l]
        r['L20'] = r.pop(x)
    return r

In [186]:
# Una vez tenemos la solución factibilizada, generamos la posibilidad de que haya una mutación en esta solución
# Se trata de intercanviar el partido del primer horario por el partido del ultimo horario
def mutar(solucion,mutacion):
    if random.random()<mutacion:
        kn = list(solucion.keys())
        d = len(kn) 
        solucion_copy = copy.deepcopy(solucion)
        solucion.update({kn[0]: solucion_copy[kn[d-1]]})
        solucion.update({kn[d-1]: solucion_copy[kn[0]]})
        return solucion
    else:
        return solucion

In [199]:
# Ejemplo cruce de una población (hijos) sin tener en cuenta la mutación en este caso, por eso ponemos 0
cruzar_ej = cruzar(poblacion_ej,0,j)
cruzar_ej

[{'V20': [['Eq3', 'Eq18']],
  'S12': ['Eq7', 'Eq20'],
  'S16': ['Eq16', 'Eq10'],
  'S18': ['Eq11', 'Eq13'],
  'S20': ['Eq15', 'Eq17'],
  'D12': ['Eq6', 'Eq1'],
  'D16': [['Eq19', 'Eq9'], ['Eq2', 'Eq14']],
  'L20': [['Eq8', 'Eq5'], ['Eq12', 'Eq4']]},
 {'V20': ['Eq8', 'Eq5'],
  'S12': [['Eq7', 'Eq20'], ['Eq16', 'Eq10']],
  'S16': [['Eq15', 'Eq17'], ['Eq6', 'Eq1']],
  'S18': ['Eq12', 'Eq4'],
  'S20': ['Eq19', 'Eq9'],
  'L20': [['Eq11', 'Eq13'], ['Eq2', 'Eq14'], ['Eq3', 'Eq18']]}]

In [204]:
# Ejemplo de una mutación (mutamos un hijo generado por el cruze). Ponemos 1 para que la probabilidad de mutación
# sea del 100%
print("Hijo antes de mutar:")
mutacion_ej

Hijo antes de mutar:


{'V20': [['Eq8', 'Eq5'], ['Eq12', 'Eq4']],
 'S12': ['Eq7', 'Eq20'],
 'S16': ['Eq16', 'Eq10'],
 'S18': ['Eq11', 'Eq13'],
 'S20': ['Eq15', 'Eq17'],
 'D12': ['Eq6', 'Eq1'],
 'D16': [['Eq19', 'Eq9'], ['Eq2', 'Eq14']],
 'L20': [['Eq3', 'Eq18']]}

In [205]:
mutacion_ej = mutar(cruzar_ej[0],1)
print("Hijo despues de mutar:")
mutacion_ej

Hijo despues de mutar:


{'V20': [['Eq3', 'Eq18']],
 'S12': ['Eq7', 'Eq20'],
 'S16': ['Eq16', 'Eq10'],
 'S18': ['Eq11', 'Eq13'],
 'S20': ['Eq15', 'Eq17'],
 'D12': ['Eq6', 'Eq1'],
 'D16': [['Eq19', 'Eq9'], ['Eq2', 'Eq14']],
 'L20': [['Eq8', 'Eq5'], ['Eq12', 'Eq4']]}

In [206]:
#Funcion de seleccion de la población. Recibe como parametro una poblacion y 
# devuelve una poblacion a la que se ha eliminado individuos poco aptos(fitness alto) y 
# para mantener una poblacion estable de N individuos
def seleccionar(poblacion, N, elitismo):
    poblacion_ordenada = sorted([ [valor(solucion), solucion] for solucion in poblacion ], key= lambda x:x[0], 
                                reverse = True)
    return [x[1] for x in poblacion_ordenada][:int(N*elitismo)]  + \
    random.sample([x[1] for x in poblacion_ordenada][int(N*elitismo):] , int(N*(1-elitismo))  )   

In [241]:
# Algoritmo genético
# N = Tamaño de la población
# mutacion = probabilidad de una mutación
# elitismo = porcion de la mejor poblacion a mantener
# generaciones = nº de generaciones a generar para finalizar 
@calcular_tiempo
def algoritmo_genetico(N,problema,mutacion,elitismo,generaciones):
    poblacion = generar_poblacion(problema,N)
    (mejor_solucion_ref, mejor_solucion_valor_ref) = evaluar_poblacion(poblacion)
    parar = False
    n=0
    
    while(parar==False):
        poblacion = cruzar(poblacion,mutacion,problema)
        poblacion = seleccionar(poblacion, N, elitismo)
        (mejor_solucion, mejor_solucion_valor) = evaluar_poblacion(poblacion)
        
        if mejor_solucion_valor > mejor_solucion_valor_ref:
            mejor_solucion_valor_ref = mejor_solucion_valor
            mejor_solucion_ref = mejor_solucion
            
        #print("Generacion #", n, "Solución:" , round(mejor_solucion_valor,3),
              #"|| Mejor Solución:",round(mejor_solucion_valor_ref,3))
            
        if n==generaciones:
            parar = True
        n +=1
        
    return mejor_solucion_ref,mejor_solucion_valor_ref       

In [209]:
# Fuerza bruta (solucion sin coincidencia de horarios)
@calcular_tiempo
def fuerza_bruta(jornada):
    mejor_solucion_valor = 0
    soluciones_posibles = list(itertools.permutations(jornada))
    for i in soluciones_posibles:
        x_sol = dic(i,key_names)
        x_val = valor(x_sol)
        if x_val > mejor_solucion_valor:
            mejor_solucion_valor = x_val
            mejor_solucion = x_sol
    return mejor_solucion, mejor_solucion_valor

In [212]:
# Jornada aleatoria
jornada = generar_jornada()
jornada

[['Eq10', 'Eq7'],
 ['Eq13', 'Eq9'],
 ['Eq2', 'Eq1'],
 ['Eq19', 'Eq8'],
 ['Eq6', 'Eq14'],
 ['Eq12', 'Eq5'],
 ['Eq4', 'Eq15'],
 ['Eq18', 'Eq16'],
 ['Eq11', 'Eq20'],
 ['Eq17', 'Eq3']]

In [213]:
# Resultado por Fuerza bruta
fuerza_bruta(jornada)

Tiempo transcurrido (en segundos): 108.85598349571228


({'V20': ['Eq19', 'Eq8'],
  'S12': ['Eq10', 'Eq7'],
  'S16': ['Eq13', 'Eq9'],
  'S18': ['Eq4', 'Eq15'],
  'S20': ['Eq2', 'Eq1'],
  'D12': ['Eq11', 'Eq20'],
  'D16': ['Eq6', 'Eq14'],
  'D18': ['Eq12', 'Eq5'],
  'D20': ['Eq17', 'Eq3'],
  'L20': ['Eq18', 'Eq16']},
 6.876)

In [226]:
# Resultado con el Algoritmo Genético 
# Tamaño población = 100
# % de mutación = 15
# % de la mejor poblacion a mantener = 40
# Nº de generaciones = 100
algoritmo_genetico(100,jornada,0.15,0.4,100)

Tiempo transcurrido (en segundos): 1.5478389263153076


({'V20': ['Eq13', 'Eq9'],
  'S12': ['Eq10', 'Eq7'],
  'S16': ['Eq6', 'Eq14'],
  'S18': ['Eq11', 'Eq20'],
  'S20': ['Eq2', 'Eq1'],
  'D12': ['Eq18', 'Eq16'],
  'D16': ['Eq4', 'Eq15'],
  'D18': ['Eq12', 'Eq5'],
  'D20': ['Eq17', 'Eq3'],
  'L20': ['Eq19', 'Eq8']},
 6.796)

In [227]:
# Resultado con el Algoritmo Genético 
# Tamaño población = 200
# % de mutación = 50
# % de la mejor poblacion a mantener = 10
# Nº de generaciones = 200
algoritmo_genetico(200,jornada,0.5,0.1,200)

Tiempo transcurrido (en segundos): 5.546178579330444


({'V20': ['Eq11', 'Eq20'],
  'S12': [['Eq13', 'Eq9']],
  'S16': ['Eq6', 'Eq14'],
  'S18': ['Eq10', 'Eq7'],
  'S20': [['Eq2', 'Eq1']],
  'D12': ['Eq18', 'Eq16'],
  'D16': [['Eq17', 'Eq3']],
  'D18': ['Eq12', 'Eq5'],
  'D20': ['Eq4', 'Eq15'],
  'L20': ['Eq19', 'Eq8']},
 6.857)

#### Conclusiones:
· Como se puede observar con fuerza bruta al ser de complejidad factorial (O(n!)), el tiempo de ejecución del algoritmo es superior al minuto y medio, con un resultado de 6.876. En canvio con algoritmos genéticos, con una población de 100 y 100 generaciones, con un tiempo de ejecución de un segundo y medio tenemos un resultado muy similar, 6.796. Y si aumentamos la población a 200 y el número de generaciones a 200, y variando un poco los parámetros de mutación y elitismo, obtenemos un mejor resultado, 6.857.<br><br>
· Tambíen podemos ver que la mejor solución con algoritmos genéticos en este caso es sin coincidencias (el algoritmo genético genera soluciones con coincidencias y sin coincidencias). <br><br>
· Nota personal: me he complicado bastante al intentar hacer todas las operaciones con diccionarios, pero me ha servido a la vez para aprender más la utilización y gestion de datos con diccionarios. Pero seria más fácil hacerlo como lo he hecho con ramificación y poda, que solo me baso en una lista de numeros.

In [260]:
# Comparación con la jornada que se ha utilizado para ramificación y poda
jornada2 =[['Eq17', 'Eq7'],
 ['Eq5', 'Eq20'],
 ['Eq12', 'Eq9'],
 ['Eq13', 'Eq2'],
 ['Eq3', 'Eq14'],
 ['Eq18', 'Eq11'],
 ['Eq10', 'Eq19'],
 ['Eq16', 'Eq1'],
 ['Eq15', 'Eq8'],
 ['Eq6', 'Eq4']]

In [262]:
# Ramificación y poda
print("Resultado: 6.746")

Resultado: 6.746


In [261]:
# Algoritmo Genético
algoritmo_genetico(100,jornada2,0.5,0.1,100)

Tiempo transcurrido (en segundos): 1.4810411930084229


({'V20': ['Eq5', 'Eq20'],
  'S12': ['Eq15', 'Eq8'],
  'S16': ['Eq12', 'Eq9'],
  'S18': ['Eq13', 'Eq2'],
  'S20': ['Eq16', 'Eq1'],
  'D12': ['Eq18', 'Eq11'],
  'D16': ['Eq3', 'Eq14'],
  'D18': ['Eq10', 'Eq19'],
  'D20': ['Eq6', 'Eq4'],
  'L20': ['Eq17', 'Eq7']},
 6.67)

> La solución del algoritmo genético esta muy cerca de la solución de ramificación y poda.